content/blog/2016/2016-08-10-rails-cd-with-pipeline.adoc

```diff
@@ -0,0 +1,358 @@
+----
+
+
+=== Preparing Jenkins
+
+[[plugins]]
+
+The cfp-app project set up, next on the list is to ensure that Jenkins itself
+is ready. Generally I suggest running the link:/changelog-stable[latest LTS] of
+Jenkins; for this demonstration I used Jenkins 2.7.1 with the following
+plugins:
+
+* link:https://wiki.jenkins-ci.org/display/JENKINS/Pipeline+Plugin[Pipeline plugin]
+* link:https://wiki.jenkins-ci.org/display/JENKINS/Brakeman+Plugin[Brakeman plugin]
+* link:https://wiki.jenkins-ci.org/display/JENKINS/CloudBees+Docker+Pipeline+Plugin[CloudBees
+  Docker Pipeline plugin]
+
+I also used the
+link:https://wiki.jenkins-ci.org/display/JENKINS/GitHub+Organization+Folder+Plugin[GitHub
+Organization Folder plugin] to automatically create pipeline items in my
+Jenkins instance; that isn't required to the demo, but it's pretty cool to see
+repositories and branches with a `Jenkinsfile` automatically show up in
+Jenkins, so I recommend it!
+
+
+In addition to the <<plugins, plugins>> listed above, I also needed at least _one_
+Jenkins agent with the link:https://docker.io[Docker] daemon installed and
+running on it. I label these agents with "docker" to make it easier to assign
+Docker-based workloads to them in the future.
+
+Any Linux-based machine with Docker installed will work, in my case I was
+provisioning on-demand agents with the
+link:https://wiki.jenkins-ci.org/display/JENKINS/Azure+Slave+plugin[Azure
+plugin] which, like the
+link:https://wiki.jenkins-ci.org/display/JENKINS/Amazon+EC2+Plugin[EC2 plugin],
+helps keep my test costs down.
+
+
+[TIP]
+--
+If you're using Amazon Web Services, you might also be interested in
+link:/blog/2016/06/10/save-costs-with-ec2-spot-fleet[this blog post] from
+earlier this year unveiling the
+link:https://wiki.jenkins-ci.org/display/JENKINS/Amazon+EC2+Fleet+Plugin[EC2
+Fleet plugin] for working with EC2 Spot Fleets.
+--
+
+
+=== Writing the Pipeline
+
+To make sense of the various things that the `Jenkinsfile` needs to do, I find
+it easier to start by simply defining the stages of my pipeline. This helps me
+think of, in broad terms, what the order of operations my pipeline should have.
+For example:
+
+[source, groovy]
+----
+/* Assign our work to an agent labelled 'docker' */
+node('docker') {
+    stage 'Prepare Container'
+    stage 'Install Gems'
+    stage 'Prepare Database'
+    stage 'Invoke Rake'
+    stage 'Security scan'
+    stage 'Deploy'
+}
+----
+
+As mentioned previously, this `Jenkinsfile` is going to rely heavily on the
+link:https://wiki.jenkins-ci.org/display/JENKINS/CloudBees+Docker+Pipeline+Plugin[CloudBees
+Docker Pipeline plugin]. The plugin provides two very important features:
+
+. Ability to execute steps _inside_ of a running Docker container
+. Ability to run a container in the "background."
+
+
+Like most Rails applications, one can effectively test the application with two
+commands: `bundle install` followed by `bundle exec rake`. I already had some
```

Comment:
I'm not 100% on what the [convention](https://en.wikipedia.org/wiki/No_true_Scotsman) is here, but for all the Rails apps I have worked on the  `:spec` task is added as a dependency to `:default`